File: src/ProxyMockTool.Core/Models/LogSort.cs
namespace ProxyMockTool.Core.Models;

/// <summary>
/// Column keys the log list can be sorted by. The names match the keys the admin UI
/// sends, so the API and the UI stay in step.
/// </summary>
public static class LogSort
{
    public const string Time = "time";
    public const string Method = "method";
    public const string Path = "path";
    public const string Protocol = "protocol";
    public const string Mode = "mode";
    public const string Status = "status";
    public const string Duration = "duration";
    /// <summary>Groups entries by proxy; only the merged (all proxies) list understands it.</summary>
    public const string Proxy = "proxy";

    /// <summary>Anything unknown falls back to the default list order: newest first.</summary>
    public static string Normalize(string? sort) => (sort ?? "").Trim().ToLowerInvariant() switch
    {
        Method => Method,
        Path => Path,
        Protocol => Protocol,
        Mode => Mode,
        Status => Status,
        Duration => Duration,
        Proxy => Proxy,
        _ => Time
    };

    /// <summary>Without an explicit sort the list is newest first.</summary>
    public static bool Descending(string? sort, bool descending) =>
        string.IsNullOrWhiteSpace(sort) || descending;
}

File: frontend/src/logSort.ts
import type { SorterResult, SortOrder } from "antd/es/table/interface";

/** Column keys the API understands, see LogSort in the backend. */
export type LogSortField =
  | "time"
  | "method"
  | "path"
  | "protocol"
  | "mode"
  | "status"
  | "duration"
  | "proxy";

export interface LogSort {
  field: LogSortField;
  descending: boolean;
}

const FIELDS: LogSortField[] = [
  "time",
  "method",
  "path",
  "protocol",
  "mode",
  "status",
  "duration",
  "proxy",
];

function isLogSortField(value: unknown): value is LogSortField {
  return typeof value === "string" && (FIELDS as string[]).includes(value);
}

/**
 * Sorter props for one column. The sort itself runs in the API so it covers the whole
 * result set instead of the page that happens to be loaded.
 */
export function logSorter(field: LogSortField, sort: LogSort | null) {
  const order: SortOrder = sort?.field === field ? (sort.descending ? "descend" : "ascend") : null;
  return { key: field, sorter: true as const, sortOrder: order };
}

/** No explicit sort means newest first, which is what the API does by default. */
export function logSortFromSorter<T>(sorter: SorterResult<T> | SorterResult<T>[]): LogSort | null {
  const result = Array.isArray(sorter) ? sorter[0] : sorter;
  const order = result?.order;
  if (order !== "ascend" && order !== "descend") {
    return null;
  }

  return isLogSortField(result?.columnKey)
    ? { field: result.columnKey, descending: order === "descend" }
    : null;
}

export function logSortParams(sort: LogSort | null): { sort?: LogSortField; descending?: boolean } {
  return sort ? { sort: sort.field, descending: sort.descending } : {};
}
